File: pyspider/monitor/monitor.py
"""
实时状态监控模块
支持爬虫状态、性能、资源监控
"""

import time
import threading
import json
import logging
from typing import Dict, List, Optional, Any, Callable
from dataclasses import dataclass, field, asdict
from datetime import datetime
from collections import deque
import os
import psutil


logger = logging.getLogger(__name__)


@dataclass
class SpiderStats:
    """爬虫统计"""
    spider_name: str
    started_at: float = 0.0
    stopped_at: float = 0.0
    pages_crawled: int = 0
    pages_failed: int = 0
    items_extracted: int = 0
    requests_made: int = 0
    bytes_downloaded: int = 0
    errors: List[Dict] = field(default_factory=list)
    
    @property
    def success_rate(self) -> float:
        total = self.pages_crawled + self.pages_failed
        return self.pages_crawled / total if total > 0 else 0
    
    @property
    def runtime(self) -> float:
        end = self.stopped_at or time.time()
        return end - self.started_at if self.started_at else 0
    
    @property
    def pages_per_second(self) -> float:
        runtime = self.runtime
        return self.pages_crawled / runtime if runtime > 0 else 0
    
    def to_dict(self) -> dict:
        return {
            **asdict(self),
            'success_rate': self.success_rate,
            'runtime': self.runtime,
            'pages_per_second': self.pages_per_second,
        }


@dataclass
class PerformanceMetrics:
    """性能指标"""
    timestamp: float = 0.0
    response_time_avg: float = 0.0
    response_time_p95: float = 0.0
    response_time_p99: float = 0.0
    requests_per_second: float = 0.0
    errors_per_second: float = 0.0
    queue_size: int = 0
    active_threads: int = 0
    memory_usage_mb: float = 0.0
    cpu_usage_percent: float = 0.0


@dataclass
class ResourceMetrics:
    """资源指标"""
    timestamp: float = 0.0
    cpu_percent: float = 0.0
    memory_percent: float = 0.0
    memory_used_mb: float = 0.0
    memory_available_mb: float = 0.0
    disk_usage_percent: float = 0.0
    network_sent_mb: float = 0.0
    network_recv_mb: float = 0.0
    open_files: int = 0
    threads_count: int = 0


class MetricsCollector:
    """指标收集器"""
    
    def __init__(self, window_size: int = 60):
        self.window_size = window_size  # 时间窗口（秒）
        self.response_times: deque = deque(maxlen=10000)
        self.request_times: deque = deque(maxlen=10000)
        self.error_times: deque = deque(maxlen=1000)
        self._lock = threading.RLock()
    
    def record_response_time(self, response_time: float):
        """记录响应时间"""
        with self._lock:
            self.response_times.append((time.time(), response_time))
    
    def record_request(self):
        """记录请求"""
        with self._lock:
            self.request_times.append(time.time())
    
    def record_error(self):
        """记录错误"""
        with self._lock:
            self.error_times.append(time.time())
    
    def get_metrics(self) -> PerformanceMetrics:
        """获取性能指标"""
        now = time.time()
        window_start = now - self.window_size
        
        with self._lock:
            # 响应时间统计
            recent_times = [
                rt for t, rt in self.response_times
                if t > window_start
            ]
            
            if recent_times:
                avg_time = sum(recent_times) / len(recent_times)
                sorted_times = sorted(recent_times)
                p95_idx = int(len(sorted_times) * 0.95)
                p99_idx = int(len(sorted_times) * 0.99)
                p95 = sorted_times[min(p95_idx, len(sorted_times) - 1)]
                p99 = sorted_times[min(p99_idx, len(sorted_times) - 1)]
            else:
                avg_time = p95 = p99 = 0.0
            
            # 请求速率
            recent_requests = sum(1 for t in self.request_times if t > window_start)
            rps = recent_requests / self.window_size
            
            # 错误速率
            recent_errors = sum(1 for t in self.error_times if t > window_start)
            eps = recent_errors / self.window_size
        
        # 系统资源
        cpu_percent = psutil.cpu_percent(interval=0.1)
        memory = psutil.virtual_memory()
        
        return PerformanceMetrics(
            timestamp=now,
            response_time_avg=avg_time,
            response_time_p95=p95,
            response_time_p99=p99,
            requests_per_second=rps,
            errors_per_second=eps,
            active_threads=threading.active_count(),
            memory_usage_mb=memory.used / 1024 / 1024,
            cpu_usage_percent=cpu_percent,
        )
    
    def clear(self):
        """清空指标"""
        with self._lock:
            self.response_times.clear()
            self.request_times.clear()
            self.error_times.clear()


class ResourceMonitor:
    """资源监控器"""
    
    def __init__(self):
        self.process = psutil.Process()
    
    def get_metrics(self) -> ResourceMetrics:
        """获取资源指标"""
        # CPU
        cpu_percent = psutil.cpu_percent(interval=0.1)
        
        # 内存
        memory = psutil.virtual_memory()
        process_memory = self.process.memory_info()
        
        # 磁盘
        disk = psutil.disk_usage('/')
        
        # 网络
        net_io = psutil.net_io_counters()
        
        # 文件句柄
        try:
            open_files = len(self.process.open_files())
        except:
            open_files = 0
        
        return ResourceMetrics(
            timestamp=time.time(),
            cpu_percent=cpu_percent,
            memory_percent=memory.percent,
            memory_used_mb=process_memory.rss / 1024 / 1024,
            memory_available_mb=memory.available / 1024 / 1024,
            disk_usage_percent=disk.percent,
            network_sent_mb=net_io.bytes_sent / 1024 / 1024,
            network_recv_mb=net_io.bytes_recv / 1024 / 1024,
            open_files=open_files,
            threads_count=self.process.num_threads(),
        )


class SpiderMonitor:
    """爬虫监控器"""
    
    def __init__(self, spider_name: str):
        self.spider_name = spider_name
        self.stats = SpiderStats(spider_name=spider_name)
        self.metrics_collector = MetricsCollector()
        self.resource_monitor = ResourceMonitor()
        
        self._callbacks: List[Callable] = []
        self._running = False
        self._monitor_thread: Optional[threading.Thread] = None
        self._lock = threading.RLock()
    
    def start(self):
        """启动监控"""
        self._running = True
        self.stats.started_at = time.time()
        
        logger.info(f"监控启动：{self.spider_name}")
    
    def stop(self):
        """停止监控"""
        self._running = False
        self.stats.stopped_at = time.time()
        
        if self._monitor_thread:
            self._monitor_thread.join(timeout=5)
        
        logger.info(f"监控停止：{self.spider_name}")
    
    def record_page_crawled(self, url: str, status_code: int, bytes_downloaded: int = 0):
        """记录页面爬取"""
        with self._lock:
            self.stats.pages_crawled += 1
            self.stats.bytes_downloaded += bytes_downloaded
            self.stats.requests_made += 1
            self.metrics_collector.record_request()
    
    def record_page_failed(self, url: str, error: str):
        """记录页面失败"""
        with self._lock:
            self.stats.pages_failed += 1
            self.stats.errors.append({
                'url': url,
                'error': error,
                'timestamp': time.time(),
            })
            self.metrics_collector.record_error()
    
    def record_item_extracted(self, count: int = 1):
        """记录数据提取"""
        with self._lock:
            self.stats.items_extracted += count
    
    def record_response_time(self, response_time: float):
        """记录响应时间"""
        self.metrics_collector.record_response_time(response_time)
    
    def add_callback(self, callback: Callable):
        """添加监控回调"""
        self._callbacks.append(callback)
    
    def get_stats(self) -> dict:
        """获取统计信息"""
        with self._lock:
            stats = self.stats.to_dict()
            metrics = self.metrics_collector.get_metrics()
            resources = self.resource_monitor.get_metrics()
            
            return {
                'spider_name': self.spider_name,
                'stats': stats,
                'performance': asdict(metrics),
                'resources': asdict(resources),
                'is_running': self._running,
            }
    
    def get_dashboard_data(self) -> dict:
        """获取仪表盘数据"""
        stats = self.get_stats()
        
        return {
            'spider_name': stats['spider_name'],
            'status': 'running' if stats['is_running'] else 'stopped',
            'pages_crawled': stats['stats']['pages_crawled'],
            'pages_failed': stats['stats']['pages_failed'],
            'success_rate': stats['stats']['success_rate'] * 100,
            'items_extracted': stats['stats']['items_extracted'],
            'bytes_downloaded': stats['stats']['bytes_downloaded'],
            'runtime_seconds': stats['stats']['runtime'],
            'pages_per_second': stats['stats']['pages_per_second'],
            'avg_response_time': stats['performance']['response_time_avg'],
            'requests_per_second': stats['performance']['requests_per_second'],
            'cpu_usage': stats['resources']['cpu_percent'],
            'memory_usage': stats['resources']['memory_used_mb'],
            'timestamp': time.time(),
        }


class MonitorCenter:
    """监控中心（管理多个爬虫）"""
    
    def __init__(self):
        self.monitors: Dict[str, SpiderMonitor] = {}
        self._lock = threading.RLock()
    
    def register(self, spider_name: str) -> SpiderMonitor:
        """注册爬虫监控"""
        with self._lock:
            if spider_name not in self.monitors:
                self.monitors[spider_name] = SpiderMonitor(spider_name)
            return self.monitors[spider_name]
    
    def unregister(self, spider_name: str):
        """注销爬虫监控"""
        with self._lock:
            if spider_name in self.monitors:
                self.monitors[spider_name].stop()
                del self.monitors[spider_name]
    
    def get_spider(self, spider_name: str) -> Optional[SpiderMonitor]:
        """获取爬虫监控器"""
        return self.monitors.get(spider_name)
    
    def get_all_stats(self) -> dict:
        """获取所有爬虫统计"""
        with self._lock:
            return {
                name: monitor.get_stats()
                for name, monitor in self.monitors.items()
            }
    
    def get_summary(self) -> dict:
        """获取摘要"""
        with self._lock:
            total_crawled = sum(m.stats.pages_crawled for m in self.monitors.values())
            total_failed = sum(m.stats.pages_failed for m in self.monitors.values())
            total_items = sum(m.stats.items_extracted for m in self.monitors.values())
            running_count = sum(1 for m in self.monitors.values() if m._running)
            
            return {
                'total_spiders': len(self.monitors),
                'running_spiders': running_count,
                'total_pages_crawled': total_crawled,
                'total_pages_failed': total_failed,
                'total_items_extracted': total_items,
                'timestamp': time.time(),
            }


# 全局监控中心
_monitor_center = MonitorCenter()


def get_monitor_center() -> MonitorCenter:
    """获取监控中心"""
    return _monitor_center


# 使用示例
if __name__ == "__main__":
    # 创建监控
    monitor = SpiderMonitor("test_spider")
    monitor.start()
    
    # 模拟爬取
    for i in range(100):
        monitor.record_page_crawled(f"https://example.com/{i}", 200, 1024)
        monitor.record_response_time(0.5 + i * 0.01)
        
        if i % 10 == 0:
            monitor.record_item_extracted(5)
    
    # 获取统计
    stats = monitor.get_stats()
    print(f"爬取页面：{stats['stats']['pages_crawled']}")
    print(f"成功率：{stats['stats']['success_rate'] * 100:.1f}%")
    print(f"平均响应时间：{stats['performance']['response_time_avg'] * 1000:.1f}ms")
    
    # 仪表盘数据
    dashboard = monitor.get_dashboard_data()
    print(f"\n仪表盘数据:")
    print(json.dumps(dashboard, indent=2))
    
    monitor.stop()

File: rustspider/src/monitor/monitor.rs
//! 实时监控模块（完整版）
//! 支持爬虫状态、性能、资源监控

use std::sync::{Arc, RwLock};
use std::time::{SystemTime, UNIX_EPOCH};
use std::collections::VecDeque;
use serde::{Serialize, Deserialize};

/// 爬虫统计
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SpiderStats {
    pub spider_name: String,
    pub started_at: f64,
    pub stopped_at: f64,
    pub pages_crawled: usize,
    pub pages_failed: usize,
    pub items_extracted: usize,
    pub requests_made: usize,
    pub bytes_downloaded: usize,
    pub errors: Vec<ErrorRecord>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ErrorRecord {
    pub url: String,
    pub error: String,
    pub timestamp: f64,
}

impl SpiderStats {
    pub fn new(spider_name: &str) -> Self {
        Self {
            spider_name: spider_name.to_string(),
            started_at: 0.0,
            stopped_at: 0.0,
            pages_crawled: 0,
            pages_failed: 0,
            items_extracted: 0,
            requests_made: 0,
            bytes_downloaded: 0,
            errors: Vec::new(),
        }
    }

    pub fn success_rate(&self) -> f64 {
        let total = self.pages_crawled + self.pages_failed;
        if total == 0 {
            0.0
        } else {
            self.pages_crawled as f64 / total as f64
        }
    }

    pub fn runtime(&self) -> f64 {
        let end = if self.stopped_at == 0.0 {
            current_timestamp()
        } else {
            self.stopped_at
        };
        end - self.started_at
    }

    pub fn pages_per_second(&self) -> f64 {
        let runtime = self.runtime();
        if runtime > 0.0 {
            self.pages_crawled as f64 / runtime
        } else {
            0.0
        }
    }
}

/// 性能指标
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PerformanceMetrics {
    pub timestamp: f64,
    pub response_time_avg: f64,
    pub response_time_p95: f64,
    pub response_time_p99: f64,
    pub requests_per_second: f64,
    pub errors_per_second: f64,
    pub queue_size: usize,
    pub active_threads: usize,
}

/// 资源指标
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResourceMetrics {
    pub timestamp: f64,
    pub cpu_percent: f64,
    pub memory_percent: f64,
    pub memory_used_mb: f64,
    pub memory_available_mb: f64,
}

/// 指标收集器
pub struct MetricsCollector {
    window_size: usize,
    response_times: Arc<RwLock<VecDeque<(f64, f64)>>>,
    request_times: Arc<RwLock<VecDeque<f64>>>,
    error_times: Arc<RwLock<VecDeque<f64>>>,
}

impl MetricsCollector {
    pub fn new(window_size: usize) -> Self {
        Self {
            window_size,
            response_times: Arc::new(RwLock::new(VecDeque::with_capacity(10000))),
            request_times: Arc::new(RwLock::new(VecDeque::with_capacity(10000))),
            error_times: Arc::new(RwLock::new(VecDeque::with_capacity(1000))),
        }
    }

    pub fn record_response_time(&self, response_time: f64) {
        let mut times = self.response_times.write().unwrap();
        let now = current_timestamp();
        times.push_back((now, response_time));
        while times.len() > 10000 {
            times.pop_front();
        }
    }

    pub fn record_request(&self) {
        let mut times = self.request_times.write().unwrap();
        times.push_back(current_timestamp());
        while times.len() > 10000 {
            times.pop_front();
        }
    }

    pub fn record_error(&self) {
        let mut times = self.error_times.write().unwrap();
        times.push_back(current_timestamp());
        while times.len() > 1000 {
            times.pop_front();
        }
    }

    pub fn get_metrics(&self) -> PerformanceMetrics {
        let now = current_timestamp();
        let window_start = now - self.window_size as f64;

        // 响应时间统计
        let times = self.response_times.read().unwrap();
        let recent_times: Vec<f64> = times
            .iter()
            .filter(|(t, _)| *t > window_start)
            .map(|(_, rt)| *rt)
            .collect();

        let (avg_time, p95, p99) = if recent_times.is_empty() {
            (0.0, 0.0, 0.0)
        } else {
            let mut sorted = recent_times.clone();
            sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());
            
            let avg = sorted.iter().sum::<f64>() / sorted.len() as f64;
            let p95_idx = (sorted.len() as f64 * 0.95) as usize;
            let p99_idx = (sorted.len() as f64 * 0.99) as usize;
            
            (
                avg,
                *sorted.get(p95_idx.min(sorted.len() - 1)).unwrap_or(&0.0),
                *sorted.get(p99_idx.min(sorted.len() - 1)).unwrap_or(&0.0),
            )
        };
        drop(times);

        // 请求速率
        let request_times = self.request_times.read().unwrap();
        let recent_requests = request_times.iter().filter(|&&t| t > window_start).count();
        let rps = recent_requests as f64 / self.window_size as f64;
        drop(request_times);

        // 错误速率
        let error_times = self.error_times.read().unwrap();
        let recent_errors = error_times.iter().filter(|&&t| t > window_start).count();
        let eps = recent_errors as f64 / self.window_size as f64;
        drop(error_times);

        PerformanceMetrics {
            timestamp: now,
            response_time_avg: avg_time,
            response_time_p95: p95,
            response_time_p99: p99,
            requests_per_second: rps,
            errors_per_second: eps,
            queue_size: 0,
            active_threads: std::thread::available_parallelism().map(|p| p.get()).unwrap_or(1),
        }
    }
}

/// 爬虫监控器
pub struct SpiderMonitor {
    pub stats: SpiderStats,
    pub metrics_collector: MetricsCollector,
    pub running: bool,
}

impl SpiderMonitor {
    pub fn new(spider_name: &str) -> Self {
        Self {
            stats: SpiderStats::new(spider_name),
            metrics_collector: MetricsCollector::new(60),
            running: false,
        }
    }

    pub fn start(&mut self) {
        self.running = true;
        self.stats.started_at = current_timestamp();
        println!("监控启动：{}", self.stats.spider_name);
    }

    pub fn stop(&mut self) {
        self.running = false;
        self.stats.stopped_at = current_timestamp();
        println!("监控停止：{}", self.stats.spider_name);
    }

    pub fn record_page_crawled(&mut self, _url: &str, _status: i32, bytes: usize) {
        self.stats.pages_crawled += 1;
        self.stats.bytes_downloaded += bytes;
        self.stats.requests_made += 1;
        self.metrics_collector.record_request();
    }

    pub fn record_page_failed(&mut self, url: &str, error: &str) {
        self.stats.pages_failed += 1;
        self.stats.errors.push(ErrorRecord {
            url: url.to_string(),
            error: error.to_string(),
            timestamp: current_timestamp(),
        });
        self.metrics_collector.record_error();
    }

    pub fn record_item_extracted(&mut self, count: usize) {
        self.stats.items_extracted += count;
    }

    pub fn record_response_time(&self, response_time: f64) {
        self.metrics_collector.record_response_time(response_time);
    }

    pub fn get_stats(&self) -> serde_json::Value {
        let metrics = self.metrics_collector.get_metrics();

        serde_json::json!({
            "spider_name": self.stats.spider_name,
            "stats": {
                "pages_crawled": self.stats.pages_crawled,
                "pages_failed": self.stats.pages_failed,
                "items_extracted": self.stats.items_extracted,
                "bytes_downloaded": self.stats.bytes_downloaded,
                "success_rate": self.stats.success_rate(),
                "runtime": self.stats.runtime(),
                "pages_per_second": self.stats.pages_per_second(),
            },
            "performance": {
                "response_time_avg": metrics.response_time_avg,
                "response_time_p95": metrics.response_time_p95,
                "response_time_p99": metrics.response_time_p99,
                "requests_per_second": metrics.requests_per_second,
                "errors_per_second": metrics.errors_per_second,
            },
            "is_running": self.running,
        })
    }

    pub fn get_dashboard_data(&self) -> serde_json::Value {
        serde_json::json!({
            "spider_name": self.stats.spider_name,
            "status": if self.running { "running" } else { "stopped" },
            "pages_crawled": self.stats.pages_crawled,
            "pages_failed": self.stats.pages_failed,
            "success_rate": self.stats.success_rate() * 100.0,
            "items_extracted": self.stats.items_extracted,
            "runtime_seconds": self.stats.runtime(),
            "pages_per_second": self.stats.pages_per_second(),
            "timestamp": current_timestamp(),
        })
    }
}

/// 监控中心
pub struct MonitorCenter {
    monitors: Arc<RwLock<std::collections::HashMap<String, Arc<RwLock<SpiderMonitor>>>>>,
}

impl MonitorCenter {
    pub fn new() -> Self {
        Self {
            monitors: Arc::new(RwLock::new(std::collections::HashMap::new())),
        }
    }

    pub fn register(&self, spider_name: &str) -> Arc<RwLock<SpiderMonitor>> {
        let mut monitors = self.monitors.write().unwrap();
        
        if !monitors.contains_key(spider_name) {
            let monitor = SpiderMonitor::new(spider_name);
            monitors.insert(spider_name.to_string(), Arc::new(RwLock::new(monitor)));
        }

        Arc::clone(monitors.get(spider_name).unwrap())
    }

    pub fn get_all_stats(&self) -> serde_json::Value {
        let monitors = self.monitors.read().unwrap();
        let stats: serde_json::Map<String, _> = monitors
            .iter()
            .map(|(name, monitor)| {
                let m = monitor.read().unwrap();
                (name.clone(), m.get_stats())
            })
            .collect();
        serde_json::Value::Object(stats)
    }

    pub fn get_summary(&self) -> serde_json::Value {
        let monitors = self.monitors.read().unwrap();
        
        let total_crawled: usize = monitors.values().map(|m| m.read().unwrap().stats.pages_crawled).sum();
        let total_failed: usize = monitors.values().map(|m| m.read().unwrap().stats.pages_failed).sum();
        let total_items: usize = monitors.values().map(|m| m.read().unwrap().stats.items_extracted).sum();
        let running_count = monitors.values().filter(|m| m.read().unwrap().running).count();

        serde_json::json!({
            "total_spiders": monitors.len(),
            "running_spiders": running_count,
            "total_pages_crawled": total_crawled,
            "total_pages_failed": total_failed,
            "total_items_extracted": total_items,
            "timestamp": current_timestamp(),
        })
    }
}

impl Default for MonitorCenter {
    fn default() -> Self {
        Self::new()
    }
}

fn current_timestamp() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs_f64()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_monitor() {
        let mut monitor = SpiderMonitor::new("test_spider");
        monitor.start();

        for i in 0..100 {
            monitor.record_page_crawled(&format!("https://example.com/{}", i), 200, 1024);
            monitor.record_response_time(0.3 + i as f64 * 0.01);
            
            if i % 10 == 0 {
                monitor.record_item_extracted(5);
            }
        }

        let stats = monitor.get_stats();
        println!("Stats: {}", stats);

        monitor.stop();
    }
}
